src/inferences/junctionTree/propagate-potentials.ts

```diff
@@ -186,12 +181,18 @@ const distributeCliquesEvidence = (network: INetwork, junctionTree: IGraph, sepS
 export default (network: INetwork, junctionTree: IGraph, cliques: IClique[], sepSets: ISepSet[], cliquesPotentials: ICliquePotentials): ICliquePotentials => {
   // Create a store for the messages passed between cliques.   Initially this store is empty because no messages have been passed.
   const messages: ICliquePotentialMessages = createMessagesByCliques(cliques)
-  const rootId: string = bestCliqueRoot(cliques).id
-
-  // Update the potentials starting at the leaf nodes and moving to the roots
-  const collectedCliquesPotentials = collectCliquesEvidence(network, junctionTree, sepSets, messages, cliquesPotentials, rootId)
-  // Update the potentials starting at the root node and moving toward the leaves
-  const distributedCliquesPotentials = distributeCliquesEvidence(network, junctionTree, sepSets, messages, collectedCliquesPotentials, rootId)
+  const ccs: string[][] = getConnectedComponents(junctionTree)
+  let potentials = cliquesPotentials
+
+  // Make each connected component of the factor graph consistent.    This is not strictly necessary for a well formed Bayes net
+  // which should have single connected component, however during incremental construction or structural learning, networks which
+  // are forests, rather than trees may occur.
+  for (const [rootId] of ccs) {
+    // Update the potentials starting at the leaf nodes and moving to the roots
+    const collectedCliquesPotentials = collectCliquesEvidence(network, junctionTree, sepSets, messages, potentials, rootId)
+    // Update the potentials starting at the root node and moving toward the leaves
+    potentials = distributeCliquesEvidence(network, junctionTree, sepSets, messages, collectedCliquesPotentials, rootId)
+  }
```

Comment:
IMHO, you could use a `reduce` here to make this calculation:
```js
  ...
  const ccs: string[][] = getConnectedComponents(junctionTree)

  // Make each connected component of the factor graph consistent.    This is not strictly necessary for a well formed Bayes net
  // which should have single connected component, however during incremental construction or structural learning, networks which
  // are forests, rather than trees may occur.
  return reduce(
    (potentials, [rootId]) => {
      // Update the potentials starting at the leaf nodes and moving to the roots
      const collectedCliquesPotentials = collectCliquesEvidence(network, junctionTree, sepSets, messages, potentials, rootId)
      // Update the potentials starting at the root node and moving toward the leaves
      return distributeCliquesEvidence(network, junctionTree, sepSets, messages, collectedCliquesPotentials, rootId)
    },
    cliquesPotentials,
    ccs,
  )
```

This way it will be shorter and you will not need to declare the `potentials` variable. 

Explaining different, all your logic and variable will be inside `reduce` :)